### DISCUSSION TITLE
How to parse/convert markdown to hast to render/display needed HTML?

### DISCUSSION THREAD

**Discussion by reypm** [syntax-tree/unist#112] [Q&A] [answered]
I am trying to turn a markdown directive defined as:
```
:::notice{node}
This is a note notice
:::
```
into this HTML:
```html
<div class="notice">
	<div class="notice-icon">
		<HiOutlineInformationCircle /> // this is coming from https://react-icons.github.io/react-icons
	<div>
	<div class="notice-content">
		This is a note notice
	</div>
</div>
```
But I am unsure how to turn such markdown into the properly hast and I will need some help/guidance. This is what I currently have:
```tsx
const processNotice = (node: any) => {
    const allowedTypes = ['warning', 'important', 'note', 'tip', 'info'];
    const type = Object.keys(node.attributes)[0];

    if (allowedTypes.indexOf(type) === -1) {
        throw new Error('Invalid notice type: ' + type);
    }

    node.children.unshift({
        type: 'containerDirective',
        children: [
            {
                type: 'svg',
                data: {
                    hName: 'svg',
                    hProperties: {
                        class: node.name + '-icon',
                        variant: type
                    }
                }
            },
            {
                type: 'paragraph',
                children: [
                    {
                        type: 'text',
                        value: 'some'
                    }
                ]
            }
        ]
    });

    // node.children.push(img)
}

export default processNotice;
```
but as you can see in the runable example: https://codesandbox.io/s/muddy-hill-gvlrvs?file=/src/app.tsx is not working :| certainly I am missing something but unsure what it is. And this one is not as easy as others where I just built an object and push it back to the node childrens.

**Comment by wooorm** [comment]
The examples in the `remark-directive` readme might help: https://github.com/remarkjs/remark-directive#examples.

As mentioned before, it is recommended to use actual types if you prefer types, instead of `any`.

Also, your code is just messy: you are not using the plugin you are making in codesandbox? You are including other code for images, but not explaining what they should do? And why do images use an `id`, set to a URL? Wouldn’t it make more sense to call it `url` or `src`?

Here’s an example I made from your code to handle container directives with a name of `image`, as you have in the markdown:

```ts
/// <reference types="remark-directive" />
import type { Image, Paragraph, Root, RootContent } from "mdast";
import { visit } from "unist-util-visit";

/**
 * Plugin to support custom directives.
 */
const myRemarkShortcodePlugin: import("unified").Plugin<
  [],
  Root
> = function () {
  return (tree, file) => {
    visit(tree, (node, index, parent) => {
      if (
        node.type === "containerDirective" &&
        typeof index === "number" &&
        parent
      ) {
        if (node.name === "image") {
          const url = node.attributes.id;

          // Require an `url`, crash otherwise.
          if (typeof url !== "string") {
            file.fail("Expected `url` attribute on image directive", node);
          }

          // Create a markdown image node.
          const img: Image = {
            type: "image",
            url,
            alt: node.attributes.alt || undefined,
            title: null
          };

          // Images cannot be directly in things.
          // They are phrasing content.
          const paragraph: Paragraph = {
            type: "paragraph",
            children: [img]
          };

          // You also don’t want to *add* it to the directive,
          // you want to *replace* the directive.
          // @ts-expect-error: content model is fine.
          const siblings: RootContent[] = parent.children;
          // Replace directive with paragraph.
          siblings.splice(index, 1, paragraph);
        }
      }
    });
  };
};

export default myRemarkShortcodePlugin;
```

**Reply by reypm** [reply]
> As mentioned before, it is recommended to use actual types if you prefer types, instead of any.

Yeah, I am trying my best there but having a hard time finding the right types to use 😞 

> Also, your code is just messy: you are not using the plugin you are making in codesandbox? 

Fixed, I did not notice that

> You are including other code for images, but not explaining what they should do? 

The code for the image is just that it will display an image from the markdown

> And why do images use an id, set to a URL? Wouldn’t it make more sense to call it url or src?

Sure, it makes sense but the shortcode (directive) developer used `id` and I cannot change that 🤦🏻 

And as for your example, I am having a hard time getting it to work in [Codesandbox](https://codesandbox.io/s/muddy-hill-gvlrvs?file=/src/customPlugins/ShortcodePlugin.ts) and locally. I have created a [repository](https://github.com/reypm/ReactMarkdownUsageExample) feel free to make a PR if you want to.

**Reply by wooorm** [reply]
copy/paste the code I wrote in `ShortcodePlugin.ts`.
Import the plugin from `App.tsx`.
Use the plugin in `remarkPlugins`.
I wrote it in codesandbox, and it worked

**Reply by reypm** [reply]
@wooorm that is exactly what I did and is not working for me in Codesanbox

**Reply by wooorm** [reply]
You have different code in ShortcodePlugin.ts, here’s it pasted in: https://codesandbox.io/s/eager-varahamihira-2tg8mw?file=/src/customPlugins/ShortcodePlugin.ts

**Reply by wooorm** [reply]
codesandbox sometimes has issues with dependencies, I also removed and added `unist-util-visit`

**Comment by reypm** [comment]
hey @wooorm many thanks for this but I think you diverge from my OP and got distracted with the image code. Can I get some ideas for the `notice` one?

**Reply by wooorm** [reply]
Yeah I solved one of your problems, hoping that you can take it from there and solve your other problem, which is similar. See also the docs, existing plugins, and spend some time trying things out yourself

**Reply by reypm** [reply]
@wooorm  Yeah that's more than clear, believe me, I have learned a LOT in this journey thanks to you and @ChristianMurphy right now I am debating myself on how to create the `mdast` structure properly since this is a little bit more complicated. This is how the node I want to process looks like (I have removed the content of the position to avoid extra-large posts):

```json
{
    "type": "containerDirective",
    "name": "notice",
    "attributes": {
        "note": ""
    },
    "children": [
        {
            "type": "paragraph",
            "children": [
                {
                    "type": "text",
                    "value": "This is a note notice!",
                    "position": {}
                }
            ],
            "position": {}
        }
    ],
    "position": {}
}
```
from the above, I guess I can reuse the `containerDirective` children without touching anything in there unless the position has to be removed or reset (I will if it is possible) but for now is not clear to me is how to embed a `Text` node inside an `Html` node since `Html` does not support children. For example, let's say I have as follows:
```tsx
    const content: Text = {
        type: 'text',
        value: 'Alpha bravo charlie.'
    };

    const div: Html = {
        type: "html",
        value: "<div>"
    };

    const paragraph: Paragraph = {
        type: "paragraph",
        children: [div, content]
    };

    // I always want to replace, this is kind of constant, right?
    const siblings: RootContent[] = parent.children;
    siblings.splice(index, 1, paragraph);

    console.log(siblings)

    const hast = toHast(node);
    const html = toHtml(hast)

    console.log(html)
```
I have tried the above code and I end up with the following:

```html
<div><p>This is a note notice!</p></div>
```
this is more or less what I am looking for, however:

- how can I add CSS attributes to the `mdast`?
- is it possible to use React components (like [react-icons](https://react-icons.github.io/react-icons/)) inside `mdast`? 

at the end, I would like to end with something like I have the idea just not sure how to deal with the icon part and the CSS classes:

```html
<div class="notice">
	<div class="notice-icon">
		<HiOutlineInformationCircle /> // this is coming from https://react-icons.github.io/react-icons
	<div>
	<div class="notice-content">
		This is a note notice
	</div>
</div>
```

**Reply by wooorm** [reply]
At this point, it’s probably a good idea to talk about what else you are using, what your entire goal is. I see React. How are you using these projects? Why use these projects? Why not MDX? What type of authors are you dealing with? etc?

**Reply by wooorm** [reply]
> is not clear to me is how to embed a Text node inside an Html node since Html does not support children.

What about the paragraph? Other things potentially in there?

**Reply by reypm** [reply]
A brief explanation of how things work on my end:

 - Authors are using Adobe Experience Manager (AEM) to write technical articles and they use Markdown most of the time
 - We do use React+NextJS: where React will of course render the content to the browser while NextJS is gathering all that information from AEM. The information is sent to us as a JSON containing Markdown inside it.
 - The Markdown written by authors does include a lot of custom directives (we call them shortcodes). Do you remember the `page-include` in which I was trying to use `async`, this `notice` and `image` are part of those custom directives.

> what your entire goal is. 

I need (or want) to use the React Markdown component to parse the JSON content, find any markdown and translate into something I can render to the UI nicely.

> I see React. How are you using these projects? 

Explained above.

> Why use these projects? 

I am open to ideas

> Why not MDX? 

Is it suitable for this scenario where the markdown is coming as a JSON string? 🤔 

> What type of authors are you dealing with? 

Persons that do not know anything technical at the code level, were trained to write text and sometimes shortcodes.

> What about the paragraph? Other things potentially in there?

What do you mean? I am not following you sorry

**Reply by wooorm** [reply]
> What do you mean? I am not following you sorry

If you look at your `console.log` call, what you posted above, there isn’t just text in the container.
There’s a paragraph.

If you would type a list in the directive, or a block quote, or emphasis, or whatever, you’d see other things in the directive, too.

Now, if you look at “what you want”, the stuff in `notice-content`, there is no paragraph there. Why not?
What if lists/blockquotes/etc are used? What happens then?

**Reply by reypm** [reply]
> Now, if you look at “what you want”, the stuff in notice-content, there is no paragraph there. Why not?
> What if lists/blockquotes/etc are used? What happens then?

I build that HTML on top of my head but It can be customized as long as it keeps the containers order and classes. So, yes, my initial idea does not have a paragraph while the resulting code (the console.log) has, which is not a problem

**Reply by wooorm** [reply]
1. Can you update a condesandbox with what you have?
2. What do you want with the `node` attribute? It’s in `:::notice{node}`
3. Are there other attributes you expect?
4. Why is `HiOutlineInformationCircle` used? Is it always that icon that is used?

**Reply by reypm** [reply]
> Can you update a condesandbox with what you have?

Sure, here: https://codesandbox.io/s/muddy-hill-gvlrvs?file=/src/app.tsx

> What do you want with the node attribute? It’s in `:::notice{node}`

The attribute will tell me what icon should I use. It maps as follows (meaning it can be any of those):

```
warning: <PiWarningLight />
important: <MdOutlineNotificationImportant />
note, info: <HiOutlineInformationCircle />
tip: <MdOutlineTipsAndUpdates />
```

> Are there other attributes you expect?

No, just that one

> Why is HiOutlineInformationCircle used? 

This is because the attribute is `note`

> Is it always that icon that is used?

No, it depends on the attribute.

Two extra things, the directive can be defined in two ways (don't ask me why 🤦🏻  is a pain) so I have to support both of them:

```
# Option 1
:::notice{note}
This is a note notice
:::

# Option 2
:::notice{title=node}
This is a notice node
:::
```
That's why my solution has this piece of code (I am open to suggestions and/or improvements 😎 ):

```tsx
  const allowedTypes = ["warning", "important", "note", "tip", "info"];
  const firstElementKey = Object.keys(node.attributes)[0];

  let currentType;

  // Require a `type` attribute, crash otherwise.
  if (Object.hasOwn(node.attributes, "title")) {
    currentType = node.attributes.title;
  } else if (allowedTypes.indexOf(firstElementKey)) {
    currentType = firstElementKey;
  } else {
    file.fail("Expected `type` attribute on notice directive", node);
  }
```

The HTML I want to output was missing something, here is the correct one:

```html
<div class="notice">
	<div class="notice-icon">
		<HiOutlineInformationCircle /> // this is coming from https://react-icons.github.io/react-icons
		Note //This is the attribute, it goes here
	<div>
	<div class="notice-content">
		This is a note notice
	</div>
</div>
```

I am not wrapping the text or the icon inside other elements but feel free to do it as needed, I need to keep the containers and the classes so the UI/UX team can play with them and get what they want

**Reply by reypm** [reply]
@wooorm Currently my approach is not working as you can see in the Codesandbox, I wonder if I need to use `remark-html` for this to get it working 🤔

**Reply by ChristianMurphy** [reply, marked as answer]
@reypm with a minimal reproducible example, aim to remove as many unrelated pieces as possible.
There are a lot of random plugins which are unrelated to your question on directives which make it harder to interpret and debug.

Back on your question of how to handle a `notice`.
Focus on how you want the `mdast` directive to translate to `hast`, using the `data` attributes in `mdast` https://github.com/syntax-tree/mdast-util-to-hast#fields-on-nodes.

```ts
import type { VFile } from "vfile";
import type { Blockquote, ListItem } from "mdast";
import type { Element } from "hast";
import type { ContainerDirective } from "mdast-util-directive";
import type { FootnoteDefinition } from "mdast-util-gfm-footnote";
import { toHast } from "mdast-util-to-hast";

/**
 * This will be called by `unist-util-visit` for each `containerDirective` in the document.
 */
function mdastContainerDirectiveVisitor(
  node: ContainerDirective,
  index: number,
  parent: Blockquote | FootnoteDefinition | ListItem | ContainerDirective,
  file: VFile
) {
  const allowedTypes = ["warning", "important", "notice", "tip", "info"];

  /**
   * Do any validation here and attach details any issue(s) on the vFile object
   */
  if (!allowedTypes.includes(node.name)) {
    file.fail("notice directive: invalid type", node);
  }

  // make sure data exists
  node.data = node.data ?? {};
  // tell remark-rehype what kind of HTML element to use for this directive
  node.data.hName = "section";
  // add attributes to the HTML element to allow you to differentiate the type
  node.data.hProperties = {
    // Using a CSS class name for example
    className: node.name
  };
  const hastChildren: Element[] = [
    // add the icon and label at the top
    {
      type: "element",
      tagName: "div",
      properties: { className: "notice-icon" },
      children: [{ type: "text", value: node.name }]
    },
    // add the content
    {
      type: "element",
      tagName: "div",
      properties: { className: "notice-content" },
      // translate directive inner content from mdast to hast
      // @ts-expect-error not all content types can be translated as inner content of element
      children: node.children.map(toHast)
    }
  ];
  node.data.hChildren = hastChildren;
}

export default mdastContainerDirectiveVisitor;
```

Runnable example: https://codesandbox.io/s/remark-directive-5lll3p?file=/src/customPlugins/Shortcodes/Notice.tsx

---

@reypm for future questions, focus on one question at a time.
Aim to reduce the question to a minimal reproducible example (https://stackoverflow.com/help/minimal-reproducible-example)
Saying a snippet of random code is "not working" doesn't give people helping you much useful context, which is frustrating for us trying to help you, aim to include the information outlined in the support guidelines https://github.com/remarkjs/.github/blob/main/support.md

**Reply by reypm** [reply]
> for future questions, focus on one question at a time. Aim to reduce the question to a minimal reproducible example 

Hi @ChristianMurphy lesson learn from now on I will try to be as much strict as possible when asking a question and thanks for the well-explained example, is more clear to me.

**Reply by reypm** [reply]
hey there @ChristianMurphy, some doubts:

1) how do I support Leafs directive as well? 🤔  The following line only support `containerDirective`:

```tsx
visit(tree, "containerDirective", (node, index, parent) => { ...
```

2) can I add custom properties like for example `variant` to the properties object while defining an Element?

```tsx
{
  type: 'element',
  tagName: 'span',
  properties: {
    className: currentType,
    variant: 'some value'
  },
  children: []
},
```
The only information I could find about the allowed properties was:
```tsx
export interface Properties {
    [PropertyName: string]: boolean | number | string | null | undefined | Array<string | number>;
}
```
any docs I should read? 🤔 

I have updated the runnable example: https://codesandbox.io/s/remark-directive-forked-hv63jx

**Reply by ChristianMurphy** [reply]
> how do I support Leafs directive as well?

There are three types of directives, `containerDirective`, leafDirective`, and `textDirective`.
https://github.com/syntax-tree/mdast-util-directive#directives

`visitor` in your example uses `Test` as its second parameter.
https://github.com/syntax-tree/unist-util-is#test
In this case an array test would allow you to capture as many or as few directive types as wanted.
```ts
visit(tree, ["containerDirective", "leafDirective", "textDirective"], (node, index, parent) => { ...
```

> can I add custom properties like for example variant to the properties object while defining an Element?

Sure, you can any pretty much anything to properties.
I'd recommend sticking with the HTML standard, and using a `data-` prefix for anything custom https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes
But that isn't a requirement imposed by hast/rehype.

**Reply by reypm** [reply]
@ChristianMurphy / @wooorm one more and last thing: while running `npm run build` the linting is failing with the following error:

```
Type error: Property 'name' does not exist on type 'ContainerDirective | LeafDirective | Blockquote | FootnoteDefinition | ListItem | HTML | ... 23 more ... | FootnoteReference'.
  Property 'name' does not exist on type 'Blockquote'.

  23 |     visit(tree, ['containerDirective', 'leafDirective', 'textDirective'], (node, index, parent) => {
  24 |       if (typeof index === "number" &&parent) {
> 25 |         switch (node.name) {
```
`node.name` is being used as follows in my custom plugin:

```tsx
const myRemarkShortcodePlugin: import("unified").Plugin<
  [],
  Root
> = function () {
  return (tree: Root, file: VFile) => {
    visit(tree, "containerDirective", (node, index, parent) => {
      switch (node.name) {
        // case "image": {
        //   processImage(node, index, parent, file);
        //   break;
        // }
        case "notice": {
          processNotice(node, index, parent, file);
          break;
        }
      }
    });
  };
};
```

is there any other property that I can use for the switch statement?

**Reply by wooorm** [reply]
This is shown in the examples, such as here: https://github.com/remarkjs/remark-directive#example-youtube.
Don’t pass a test to `unist-util-visit` when TS doesn’t understand it. Do it yourself with an if statement to check node types, and then TS understands it. I’ve also shown this here: https://github.com/orgs/syntax-tree/discussions/112#discussioncomment-6699560

**Reply by ChristianMurphy** [reply]
TL;DR the complexity of the types in https://github.com/syntax-tree/unist-util-visit/blob/bc59f3bb488a71a1aab3da094306022eeecabda8/lib/index.js#L31-L63
made it infeasible to infer a list of conditions (folks are welcome to research more how to support it, without significantly slowing or crashing typescript).

You can either, explicitly type the parameters of the visit function.
Or, consider replacing the array with one of the approaches outlined in https://unifiedjs.com/learn/recipe/narrow-node-typescript/

An example of the latter:

```ts
import type { Root } from "mdast";
import type { Directive } from "mdast-util-directive";
import type { VFile } from "vfile";
import { visit } from "unist-util-visit";
import { convert } from "unist-util-is";
import processNotice from "./Shortcodes/Notice";

/**
 * Plugin to support custom directives.
 */
const myRemarkShortcodePlugin: import("unified").Plugin<
  [],
  Root
> = function () {
  return (tree: Root, file: VFile) => {
    visit(
      tree,
      convert<Directive>([
        "containerDirective",
        "leafDirective",
        "textDirective"
      ]),
      (node, index, parent) => {
        switch (node.name) {
          case "notice": {
            processNotice(node, index, parent, file);
            break;
          }
        }
      }
    );
  };
};

export default myRemarkShortcodePlugin;
```
